executor/aggfuncs/func_percentile.go

```diff
@@ -0,0 +1,401 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggfuncs
+
+import (
+	"math"
+	"sort"
+	"unsafe"
+
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/selection"
+)
+
+const (
+	// DefPartialResult4PercentileIntSize is the size of partialResult4PercentileInt
+	DefPartialResult4PercentileIntSize = int64(unsafe.Sizeof(partialResult4PercentileInt{}))
+	// DefPartialResult4PercentileRealSize is the size of partialResult4PercentileReal
+	DefPartialResult4PercentileRealSize = int64(unsafe.Sizeof(partialResult4PercentileReal{}))
+	// DefPartialResult4PercentileDecimalSize is the size of partialResult4PercentileDecimal
+	DefPartialResult4PercentileDecimalSize = int64(unsafe.Sizeof(partialResult4PercentileDecimal{}))
+	// DefPartialResult4PercentileTimeSize is the size of partialResult4PercentileTime
+	DefPartialResult4PercentileTimeSize = int64(unsafe.Sizeof(partialResult4PercentileTime{}))
+)
+
+func percentile(data sort.Interface, percent int) int {
+	// Ordinal rank k = Ceil(P / 100 * N)
+	k := int(math.Ceil(float64(data.Len()) / 100 * float64(percent)))
+	return selection.Select(data, k)
+}
+
+type basePercentile struct {
+	percent int
+
+	baseAggFunc
+}
+
+func (e *basePercentile) AllocPartialResult() (pr PartialResult, memDelta int64) {
+	return
+}
+
+func (e *basePercentile) ResetPartialResult(pr PartialResult) {
+	return
+}
+
+func (e *basePercentile) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup []chunk.Row, pr PartialResult) (memDelta int64, err error) {
+	return
+}
+
+func (e *basePercentile) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	chk.AppendNull(e.ordinal)
+	return nil
+}
+
+type intArray []int64
+
+func (a intArray) Len() int           { return len(a) }
+func (a intArray) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a intArray) Less(i, j int) bool { return a[i] < a[j] }
+
+type partialResult4PercentileInt struct {
+	data intArray
+}
+
+func (p *partialResult4PercentileInt) MemSize() int64 {
+	return DefPartialResult4PercentileIntSize + int64(len(p.data))*DefInt64Size
+}
+
+type realArray []float64
+
+func (a realArray) Len() int           { return len(a) }
+func (a realArray) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a realArray) Less(i, j int) bool { return a[i] < a[j] }
+
+type partialResult4PercentileReal struct {
```

Comment:
We create a interface like
```go
type percentileArray interface {
	Len() int
	Swap(i, j int)
	Less(i, j int) bool
	MemSize() int64
}
```
Then there is no need to define so many `partialResult4Percentile` for each type.

Reply:
You are right, I'll fix it.